server/src/main/java/org/opensearch/common/metrics/ResourceTracker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.metrics;
+
+import com.sun.management.ThreadMXBean;
+import org.opensearch.common.SuppressForbidden;
+
+import java.lang.management.ManagementFactory;
+
+/*
+    Simple tracker for CPU consumption and memory allocated by the current thread
+*/
+@SuppressForbidden(reason = "ThreadMXBean enables tracking resource consumption by a thread. "
+    + "It is platform dependent and i am not aware of an alternate mechanism to extract this info")
+public class ResourceTracker {
+
+    static ThreadMXBean threadMXBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+    long startingAllocatedBytes;
+    long startingCPUTime;
+    long memoryAllocated;
+    long cpuTime;
+
+    public ResourceTracker() {
+        reset();
+    }
+
+    /**
+     * Takes current snapshot of resource usage by thread since the creation of this object
+     */
+    public void updateMetrics() {
+        this.memoryAllocated = threadMXBean.getThreadAllocatedBytes(Thread.currentThread().getId()) - startingAllocatedBytes;
+        this.cpuTime = threadMXBean.getCurrentThreadCpuTime() - startingCPUTime;
+    }
+
+    /**
+     * Wipes out local state and resets to a clean tracker
+     */
+    public void reset() {
+        this.startingCPUTime = threadMXBean.getCurrentThreadCpuTime();
+        this.startingAllocatedBytes = threadMXBean.getThreadAllocatedBytes(Thread.currentThread().getId());
```

Comment:
I think it would be safer to consult if  thread memory allocation measurement is supported (otherwise `UnsupportedOperationException`) and enabled [1]:

```
if (threadMXBean.isThreadAllocatedMemorySupported() && threadMXBean.isThreadAllocatedMemoryEnabled {
    this.startingAllocatedBytes = threadMXBean.getThreadAllocatedBytes(Thread.currentThread().getId());
}
```
[1] https://docs.oracle.com/javase/8/docs/jre/api/management/extension/com/sun/management/ThreadMXBean.html